### ISSUE TITLE
Remove the queues between the stages

### ISSUE THREAD

**Issue by bmbouter** [pulp/pulpcore#2377]
Recently the queue sizes got changed to 1 from 1000, and the question came up why do we have queues at all?

## Original Idea

The original idea was to create backpressure on the Stages, but in practice it just created a lot more objects in-memory while not actually reducing runtime.

## New Idea

What if we removed the queues altogether and made it so that we only store declarative objects in the batcher that belongs to the stage.

**Comment by mdellweg** [comment]
Maybe we should experiment with queue sizes of 1, 5, 100 and 1000 and use the results for a decision.

**Comment by bmbouter** [comment]
My pitch here is to remove it from the architecture entirely. To me, that's the only option to further reduce memory and runtime together, but yes to more testing to be sure. If we had a PoC of the queue removal we could do an evaluation like that.

**Comment by dralley** [comment]
I doubt it would have much mem / runtime impact, but I suppose the point is that if the queues were never actually accomplishing the thing we thought they were, then they're pure downside regardless of if it's 1, 5, 10, 100, 1000.

**Comment by bmbouter** [comment]
I agree the impact won't be huge, but it's got to be slightly better than pushing them through an entirely unnecessary data structure. Due to the likely small impact though, I think this should not a medium or low priority.

**Comment by mdellweg** [comment]
My point is that we have stages that batch their incoming flow, and these won't benefit from extra piling up in a different data structure, but we also have stages that process items one by one. The queues between them make their operations less rigid. What is unclear to me if that is beneficial to the runtime in any way.

**Comment by bmbouter** [comment]
If you can run some tests and share the results that would be cool. When I was running it, I was doing an EL7 sync on_demand.

**Comment by bmbouter** [comment]
I don't think this needs to be on the 3.20 blockers list. If we still get to it great, but it's not a blocker par-se.

**Comment by dralley** [comment]
Some other use cases to consider, perhaps the "pipeline" machinery ought to be a library available to plugins where plugins can use it at a more granular level rather than passing off control entirely.

https://github.com/pulp/pulp_rpm/issues/2432#issuecomment-1194171295
